### DISCUSSION TITLE
MSBuild support / deep MSVC integration for ISPC

### DISCUSSION THREAD

**Discussion by ishani** [ispc/ispc#1958] [Show and tell]
I've just checked in my integration of ISPC into Visual Studio via MSBuild extension - [Tether-ISPC](https://github.com/ishani/Tether-ISPC) . This means you get proper `.ispc` / `.isph` change tracking and complete Properties management in the IDE. 

![vs_props](https://user-images.githubusercontent.com/142049/103157622-c2053400-47ac-11eb-8b42-cdf67a430255.png)

Trivial to integrate into existing vcproj and no more having to do custom build steps / running ISPC separately.

It also has a bunch of examples, tests and benchmarks in there, [Premake 5](https://premake.github.io/download.html) support and an experimental pipeline for compile-and-deploy on Raspberry Pi using ISPC's NEON support

![vs_output_pi](https://user-images.githubusercontent.com/142049/103157672-5f606800-47ad-11eb-86d3-b8467ba82d62.png)

If anyone fancies giving it a look and let me know any feedback, would appreciate it!

**Comment by pbrubaker** [comment]
Hey Harry,

This is AWESOME!  I have just checked it out a little bit, but it appears
as though this checks off a lot of things on my todo list.  I wrote
the vscode plugin and I have been working on language integration of ISPC
with VS for a while now.  One of the last things I needed to do before
releasing was integrate with the build system.  My goal is to have the vsix
package install ISPC MSBuild support as a "Build Customization."  Once this
basic functionality is done I plan on releasing the source to both as part
of the ISPC repo.  It would be great to make this a part of the official
plugin.  What do you think?

Cheers,

Pete

**Reply by ishani** [reply]
Would **love** to have proper language integration in VS so yeah, absolutely onboard with getting this integrated as part of that. Although I appreciate your vscode plugin a great deal, I'd love to not have VS and VSC to hand to edit my ISPC source :)

Am happy to let you take whatever you need from Tether, or let me know if there's something I can do to help prep. The main change I think might be required is to improve the "where is ISPC.exe installed" section of the build script; either perhaps by environment variable, or if the VSIX installs ISPC too, then perhaps relative to that

The rest of it is pretty well battle-tested, although if it's likely to go to a wider audience I'll go maybe make some test projects to ensure my coverage of all the little switches and options actually function as expected.

**Reply by degracode** [reply]
How did you get on with this @pbrubaker ?
There's been some chatter about VS project support in the cmake repo - https://gitlab.kitware.com/cmake/cmake/-/issues/22596#note_1610265

**Reply by pbrubaker** [reply]
@degracode Very little progress, just researching, collecting resources and taking notes.  I had commented on that thread there and I was hoping they'd be able to take care of it but it seems like we'll have to have a proper MSBuild task for that.

I have pivoted to making an ISPC cmake file, and creating a couple github templates to help people get started quickly.  I decided doing that will meet almost all my goals and get done quicker.

I still want to do proper Visual Studio integration, but with the layoffs (myself included) I lost the person that was working on the tools.  As I understand it, the vscode extension repo is going through legal review now.  Then we will be releasing a large update to that, and I can shift focus back to Visual Studio.

**Reply by degracode** [reply]
My condolences on your job situation, it's been a difficult few years.

Also only just realised that you were in the other thread already 😂.

Is there anything I can help with? I'm happy to put some evenings in, but I'm not sure where the best place to start with this would be.

**Reply by pbrubaker** [reply]
Hey @degracode.  Thanks for your kind thoughts.  I'll be fine.  Hiring is picking up in the new year, there are many more opportunities.

> Is there anything I can help with? I'm happy to put some evenings in, but I'm not sure where the best place to start with this would be.

Yes, I do have an idea!  I would love if someone could dig into how the CUDA and Intel FORTRAN plugins/extensions for Visual Studio work.  Maybe take a look and share some notes here?   What is the work flow for the programmer?  Do they integrate with existing project templates?  Can you add them to an existing C++ project?  Also, maybe look into general extension development?  Basically define a bare bones skeleton plan to follow when we are able to tackle it.  What do you think?

**Reply by degracode** [reply]
Sounds like a good starting point! I'll have a look at FORTRAN...
https://docs.google.com/document/d/1JNZsefaH5L2drmzNrm5-LOfPUZWu_sHRzSRVRNAsFE4/edit?tab=t.0

**Reply by degracode** [reply]
FORTRAN was interesting as a complete commercial-class language integration - it looks like there's a lot of work that's gone into a new project type for it, with suites of template projects. Unfortunately VS doesn't understand what to do if you add a .f90 file to a C++ project; it doesn't participate in the build.

I then went looking for open source examples and found [Visual D](https://github.com/dlang/visuald), which does appears to properly support mixed language in a single project; you can add D files to a vcpkg, and Visual Studio understands to use the D compiler on them (though I've not actually produced a working exe yet). Which is more like what Harry'd built with Tether.

**Reply by pbrubaker** [reply]
>it looks like there's a lot of work that's gone into a new project type for it,

> Unfortunately VS doesn't understand what to do if you add a .f90 file to a C++ project; it doesn't participate in the build.

Thanks for looking into this @degracode.  This is what I thought was most likely and what I was hoping wasn't the case.  Essentially we are going to have to build a new project type for ISPC but also incorporates everything needed from a default C++ template.  So if we want proper integration and user experience we will have to do something similar.  

Doing this will have rippling effects in the developer ecosystem as well.

https://gitlab.kitware.com/cmake/cmake/-/issues/22596

Your experiments are encouraging, do you want to dig into this some more?

**Reply by pbrubaker** [reply]
@ishani I know this is a long time ago but would you be willing to give this another look?

**Reply by ishani** [reply]
Hey @pbrubaker  - sorry for the late reply, have been very busy elsewhere.

What would you need, exactly? I've done a quick update today to Tether to support ISPC 1.26 but not really tested or bedded it down much yet.

I'm more than happy to polish it up a bit further and you're very welcome to bundle the MSBuild configs however you like. I had dreamed of doing a more substantial VS integration but can't spare the time to head off down that rabbit hole currently :(
